File: src/py/207-course-schedule/main.py
class Solution:
    @classmethod 
    def naive(cls, numCourses, prerequisites):
        cls.adjls = {}
        for take, require in prerequisites:
            cls.adjls[take] = cls.adjls.get(take, []) + [require]
        print(cls.adjls)
        cls.cache = {}
        def dfs(i, visited):
            if len(cls.adjls.get(i, [])) == 0:
                print(visited|set([i]))
                cls.cache[i] = True
                return True
            if i in visited:
                return False
            if i in cls.cache:
                print(visited|set([i]))
                return cls.cache[i]
            for child in cls.adjls[i]:
                isFinish = dfs(child, visited|set([i]))
                cls.cache[child] = isFinish
                if not isFinish:
                    return False
            return True

        for i in range(numCourses):
            isFinish = dfs(i, set())
            cls.cache[i] = isFinish
            if not isFinish:
                return False
        print(cls.cache)
        return True

if __name__ == "__main__":
    inputs = [
        (2, [[1,0]], True),
        (2, [[1,0],[0,1]], False),
        (6, [[5,4],[5,3],[3,1],[4,2],[1,0],[2,0]], True),
        (7, [[1,4],[3,4],[5,6],[5,1],[2,1],[2,3],[0,5],[0,2]], True),
        (7, [[1,4],[3,4],[5,6],[5,1],[2,1],[2,3],[0,5],[2,0]], True),
        (4, [[5,1],[1,2],[2,0],[0,5]], False),
    ]
    for numCourses, prerequisites, expected in inputs:
        print(Solution.naive(numCourses,prerequisites))
        print("------")

File: src/java/207-course-schedule/main.java
import java.util.Hashtable;
import java.util.ArrayList;
import java.util.HashSet;

class Solution {
    private Hashtable<Integer, List<Integer>> adjls;
    private Hashtable<Integer, Boolean> cache;
    private void convert_to_adjacent_list(int[][] prerequisites) {
        this.adjls = new Hashtable<>();
        for (int[] prerequisite: prerequisites) {
            List<Integer> requires = adjls.getOrDefault(prerequisite[0], new ArrayList<Integer>());
            requires.add(prerequisite[1]);
            this.adjls.put(prerequisite[0], requires);
        }
    }
    private boolean dfs(int i, HashSet<Integer> visited) {
        if (this.adjls.getOrDefault(i, new ArrayList<Integer>()).size() == 0) {
            this.cache.put(i, true);
            return true;
        }
        if (visited.contains(i)) 
            return false;
        HashSet newVisited = new HashSet();
        newVisited = (HashSet)visited.clone();
        newVisited.add(i);
        if (this.cache.containsKey(i))
            return this.cache.get(i);
        for (int child : this.adjls.getOrDefault(i, new ArrayList<Integer>())) {
            boolean isFinish = this.dfs(child, newVisited);
            this.cache.put(child, isFinish);
            if (! isFinish)
                return false;
        }
        return true;
    }
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        this.convert_to_adjacent_list(prerequisites);
        this.cache = new Hashtable<>();
        for (int i=0; i<numCourses; i++){
            boolean isFinish = this.dfs(i, new HashSet<Integer>());
            this.cache.put(i, isFinish);
            if (! isFinish)
                return false;
        }
        return true;
    }
}